### PR TITLE
Add [Compile] contextual menu on specific file extension

### PR DESCRIPTION
**Pull request by hyunsik-yoon**
This adds [Compile] contextual menu when user right-clicks
compilable model file, e.g., *.tflite.

Related issue #331 (to show compile screen)

ONE-vscode-DCO-1.0-Signed-off-by: Hyun Sik Yoon <hyunsik.yoon.1024@gmail.com>

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -35,6 +35,12 @@
       ]
     },
     "commands": [
+      {
+        "command": "onevscode.show-compile-webview",
+        "title": "Compile",
+        "category": "ONE",
+        "_comment": "Call this when compiling compilable model(e.g., *.tflite, *.onnx, etc.)"
```
Comment (hyunsik-yoon) [inline package.json:42]: Note: `_comment` was not defined by vscode nor nodejs. I used this non-standard key name to add comment. To distinguish this with standard key names, I put underbar.

File: package.json [hunk 2]
```diff
@@ -67,6 +73,12 @@
       }
     ],
     "menus": {
+      "explorer/context": [
+        {
+          "command": "onevscode.show-compile-webview",
+          "when": "resourceExtname in onevscode.compilableExtList || resourcePath in onevscode.compilableDirList"
```
Comment (hyunsik-yoon) [inline package.json:79]: ref: https://code.visualstudio.com/api/references/when-clause-contexts#in-conditional-operator

File: src/Support/CompilableFile.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    case B) When vscode runs on Linux and workspace is on Linux
+    case C) When vscode and workspace is in Windows
+    case D) When vscode and workspace is in Linux
+
+    vscode context values in the above three cases are:
+
+       context name   |    case A        |     case B     |     case C     |     case D
+       ---------------|------------------|----------------|----------------|----------------
+       isLinux        |      false       |       ?        |       ?        |       ?
+       isWindows      |       true       |       ?        |       ?        |       ?
```
Comment (llFreetimell) [inline src/Support/CompilableFile.ts:49]: `isLinux` and `isWindows` are based on which environment? VSCode or workspace?
If they are based on only one thing, naming like `isLinuxWorkspace` or `isLinuxUI` seems better :)

File: src/Support/CompilableFile.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+    TODO Investigate case B and case C and fill the above table
+
+    TODO iterate directory in workspace and search dirs that contains Keras model and saved model
+
+    TODO Check if this instance of extension runs in case A, B or C
+
+    In case of case A), call
+
+        dirList.push(`\\tmp//foo`);
```
Comment (llFreetimell) [inline src/Support/CompilableFile.ts:64]: Maybe?
```suggestion
        dirList.push(`\\tmp\\foo`);
```

File: src/Support/CompilableFile.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+       context name   |    case A        |     case B     |     case C     |     case D
+       ---------------|------------------|----------------|----------------|----------------
+       isLinux        |      false       |       ?        |       ?        |       ?
+       isWindows      |       true       |       ?        |       ?        |       ?
+       resourceScheme | "vscode-remote"  |       ?        |       ?        |       ?
+       resourcePath   | "\\tmp\\foo"     |       ?        |       ?        |       ?
+
+    Case A) was investigated to check possibility in draft. As you can see, path separator is in
+    Windows format while path exists in Linux.
```
Comment (llFreetimell) [inline src/Support/CompilableFile.ts:54]: I think case A ~ case D have exactly same reason to be investigated.
How about remove this comment and add "what this function want to do" at the top of this function?

File: src/extension.ts [hunk 6]
```diff
@@ -24,11 +24,26 @@ import {CodelensProvider} from './Editor/CodelensProvider';
 import {HoverProvider} from './Editor/HoverProvider';
 import {Jsontracer} from './Jsontracer';
 import {Project} from './Project';
+import {CompilableFile} from './Support/CompilableFile';
 import {Utils} from './Utils';
 
 export function activate(context: vscode.ExtensionContext) {
   console.log('one-vscode activate OK');
 
+  // Thses context is used to show "Compile" menu in Explorer view
```
Comment (llFreetimell) [inline src/extension.ts:33]:
```suggestion
  // These context is used to show "Compile" menu in Explorer view
```

File: src/Support/CompilableFile.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as vscode from 'vscode';
+
+export class CompilableFile {
```
Comment (YongseopKim) [inline src/Support/CompilableFile.ts:19]: Hmm, I'm not sure this name.

Comment (YongseopKim) [reply src/Support/CompilableFile.ts:19]: If we use the name, how about this?

```ts
class CompilableExts {
 exts = ['pb', 'tflite'];
}

// activate() in extension.ts
// ...
vscode.commands.executeCommand('setContext', 'onevscode.compilableExtList', CompilableExts());
```

File: src/Support/CompilableFile.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
+/*
```
Comment (YongseopKim) [inline src/Support/CompilableFile.ts:1]: And this file is located in `Support/`. If we introduce dir `Support`, what classes and functions would be in? I'm not sure.

File: src/extension.ts [hunk 9]
```diff
@@ -24,11 +24,26 @@ import {CodelensProvider} from './Editor/CodelensProvider';
...
+  //
+  // 1. When a file is right-clicked (e.g., .pb, .tflite, etc)
+  // 2. When a dir is right-clicked (e.g., Keras model or saved model)
+  CompilableFile.setFileExtContext('onevscode.compilableExtList');
```
Comment (YongseopKim) [inline src/extension.ts:37]: To me, it seems that `CompilableFile` is there and its file ext context is `onevscode.compilableExtList`.

File: package.json (hunk 1 above)
Comment (seanshpark) [reply package.json:42]: Is there any (doc) file that describes this? if not, how about adding this style to some doc file?

Comment (seanshpark) [reply package.json:42]: (please merge this without my approval... left just a comment not a review opinion)

Comment (hyunsik-yoon) [reply package.json:42]: Good idea. I will think about this and create a new issue.

Comment (hyunsik-yoon) [reply package.json:42]: #420 was created.

File: src/Support/CompilableFile.ts (hunk 8 above)
Comment (hyunsik-yoon) [reply src/Support/CompilableFile.ts:1]: Maybe I introduce too many change with PR. Let me undo the introduction of this directory.

File: src/Support/CompilableFile.ts (hunk 7 above)
Comment (hyunsik-yoon) [reply src/Support/CompilableFile.ts:19]: > Hmm, I'm not sure this name.

Could you let me know the reason why you  don't like the name `CompilableFile`?

**Comment by hyunsik-yoon** [comment]
For reviewers,

For https://github.com/Samsung/ONE-vscode/pull/419#discussion_r836183974, I moved code from `Support` directory into `extensions.ts` and tries to shorten the code.
- 1. previous `class` was changed to `function`
- 2. previous comment was moved to issue comment in github. I believe that this is also good way to provide information with less reviewing effort.

PTAL.

**Review by mhs4670go — APPROVED** [review]
LGTM

**Review by YongseopKim — APPROVED** [review]
LGTM 👍

**Review by llFreetimell — APPROVED** [review]
LGTM

**Pull request merged**